source/compiler/qsc_qasm/src/compiler.rs

```diff
@@ -184,14 +186,37 @@ impl QasmCompiler {
         for pragma in &program.pragmas {
             self.compile_pragma_stmt(pragma);
         }
+
         self.compile_stmts(&program.statements);
         let (package, signature) = match program_ty {
             ProgramType::File => self.build_file(),
             ProgramType::Operation => self.build_operation(),
             ProgramType::Fragments => (self.build_fragments(), None),
         };
 
-        QasmCompileUnit::new(self.source_map, self.errors, package, signature)
+        let target_profile = self.get_profile();
+        QasmCompileUnit::new(
+            self.source_map,
+            self.errors,
+            package,
+            signature,
+            target_profile,
+        )
+    }
+
+    /// Gets the profile for compilation from the first profile
+    /// pragma if present, otherwise default to `Unrestricted`.
+    fn get_profile(&self) -> Profile {
+        self.pragma_config.pragmas
+            .iter()
+            .find_map(|(kind, value)|
+                if matches!(kind, PragmaKind::QdkQirProfile) {
+                    Some(Profile::from_str(value).expect("Invalid profile pragma; only a valid profile should be store in pragma_config."))
+                } else {
+                    None
+                }
+            )
+            .unwrap_or(Profile::Unrestricted)
     }
```

Comment:
Shouldn't iterate, this is a hash lookup.
```rust
self
    .pragma_config
    .pragmas
    .get(&PragmaKind::QdkQirProfile)
    .map(...)
    .unwrap_or(Profile::Unrestricted)
```

Reply:
done